### PR TITLE
Fix NRE in DeclarationNameCompletionProvider

### PR DESCRIPTION
**Pull request by genlu**
Noticed this while debugging something else.

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs [hunk 1]
```diff
@@ -71,7 +71,7 @@ public override async Task ProvideCompletionsAsync(CompletionContext completionC
                 using var _ = ArrayBuilder<(string name, SymbolKind kind)>.GetInstance(out var result);
 
                 // Suggest names from existing overloads.
-                if (nameInfo.PossibleSymbolKinds.Any(k => k.SymbolKind == SymbolKind.Parameter))
+                if (!nameInfo.PossibleSymbolKinds.IsDefaultOrEmpty && nameInfo.PossibleSymbolKinds.Any(k => k.SymbolKind == SymbolKind.Parameter))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs:74]:
```suggestion
                if (nameInfo.PossibleSymbolKinds.NullToEmpty().Any(k => k.SymbolKind == SymbolKind.Parameter))
```

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs (hunk 1 above)
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs:74]: ❗ We should instead required the constructor to initialize `PossibleSyntaxKinds` to a non-default value.

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs (hunk 1 above)
Comment (genlu) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs]: Fixed

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 2]
```diff
@@ -72,7 +72,12 @@ internal static async Task<NameDeclarationInfo> GetDeclarationInfoAsync(Document
                     return result;
                 }
 
-                return default;
+                return new(
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs:75]:
```suggestion
                return new NameDeclarationInfo(
```

**Review by sharwell — APPROVED** [review]

**Comment by genlu** [comment]
Oh wow, this is bad. It seems the provider is actually relying on the exception to have the right behavior.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 3]
```diff
@@ -636,7 +636,7 @@ class C
                               showCompletionInArgumentLists:=showCompletionInArgumentLists, languageVersion:=LanguageVersion.Preview)
 
                 state.SendTypeChars("w")
-                Await state.AssertSelectedCompletionItem(displayText:="with", isHardSelected:=False)
+                Await state.AssertSelectedCompletionItem(displayText:="with", isHardSelected:=True)
```
Comment (genlu) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:639]: Note the behavior change here. I believe previous behavior is wrong.

**Pull request merged**